src/core/rangeset.js

```diff
@@ -1,67 +1,61 @@
-var assert = require('assert');
-var lodash = require('lodash');
-
-function RangeSet() {
-  this._ranges = [ ];
-};
-
-/**
- * Add a ledger range
- *
- * @param {Number|String} range string (n-n2,n3-n4)
- */
-
-RangeSet.prototype.add = function(range) {
-  assert(typeof range !== 'number' || !isNaN(range), 'Ledger range malformed');
-
-  range = String(range).split(',');
-
-  if (range.length > 1) {
-    return range.forEach(this.add, this);
-  }
-
-  range = range[0].split('-').map(Number);
-
-  var lRange = {
-    start: range[0],
-    end: range[range.length === 1 ? 0 : 1]
-  };
-
-  // Comparisons on NaN should be falsy
-  assert(lRange.start <= lRange.end, 'Ledger range malformed');
-
-  var insertionPoint = lodash.sortedIndex(this._ranges, lRange, function(r) {
-    return r.start;
+/* @flow */
+'use strict';
+const _ = require('lodash');
+const assert = require('assert');
+const ranges = Symbol();
+
+function mergeIntervals(intervals: Array<[number, number]>) {
+  const stack = [[-Infinity, -Infinity]];
+  _.forEach(_.sortBy(intervals, x => x[0]), interval => {
+    const lastInterval = stack.pop();
+    if (interval[0] <= lastInterval[1] + 1) {
+      stack.push([lastInterval[0], Math.max(interval[1], lastInterval[1])]);
+    } else {
+      stack.push(lastInterval);
+      stack.push(interval);
+    }
   });
+  return stack.slice(1);
+}
 
-  this._ranges.splice(insertionPoint, 0, lRange);
-};
+class RangeSet {
+  constructor() {
+    this.reset();
+  }
 
+  reset() {
+    this[ranges] = [];
+  }
 
-/*
- * Check presence of ledger in range
- *
- * @param {Number|String} ledger
- * @return Boolean
- */
+  serialize() {
+    return this[ranges].map(range =>
+      range[0].toString() + '-' + range[1].toString()).join(',');
+  }
 
-RangeSet.prototype.has =
-RangeSet.prototype.contains = function(ledger) {
-  assert(ledger != null && !isNaN(ledger), 'Ledger must be a number');
+  addRange(start: number, end: number) {
+    assert(start <= end, 'invalid range');
+    this[ranges] = mergeIntervals(this[ranges].concat([[start, end]]));
+  }
 
-  ledger = Number(ledger);
+  addValue(value: number) {
+    this.addRange(value, value);
+  }
 
-  return this._ranges.some(function(r) {
-    return ledger >= r.start && ledger <= r.end;
-  });
-};
+  addParsedRanges(rangesString: string) {
```

Comment:
An unparsed range would be a string, a parsed range is a range. The other name I considered was `parseRangesAndAdd`, but I thought `addParsedRanges` sounded simpler.